runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/HttpBody.kt

```diff
@@ -144,6 +147,47 @@ public fun SdkSource.toHttpBody(contentLength: Long? = null): HttpBody =
         override fun readFrom(): SdkSource = this@toHttpBody
     }
 
+/**
+ * Convert an [HttpBody.SourceContent] or [HttpBody.ChannelContent] to a body with a [HashingSource] or [HashingByteReadChannel], respectively.
+ * @param hashFunction the hash function to wrap the body with
+ * @param contentLength the total content length of the source, if known
+ */
+@InternalApi
+public fun HttpBody.toHashingBody(hashFunction: HashFunction, contentLength: Long?): HttpBody = when (this) {
+    is HttpBody.SourceContent ->
+        HashingSource(
+            hashFunction,
+            readFrom(),
+        ).toHttpBody(contentLength)
+    is HttpBody.ChannelContent -> HashingByteReadChannel(
+        hashFunction,
+        readFrom(),
+    ).toHttpBody(contentLength)
+    else -> throw ClientException("HttpBody type is not supported")
+}
+
+/**
+ * Convert an [HttpBody.SourceContent] with a [HashingSource] or [HttpBody.ChannelContent] with a [HashingByteReadChannel]
+ * to a body which will use the digest to complete a [CompletableDeferred]
+ * @param completableDeferred the [CompletableDeferred] to complete once the underlying hashing body is exhausted
+ * @param contentLength the total content length of the source, if known
+ */
+@InternalApi
+public fun HttpBody.toCompletingBody(completableDeferred: CompletableDeferred<String>, contentLength: Long?): HttpBody =
```

Comment:
**comment**: Same as before this doesn't feel general purpose enough and I'd maybe consider moving some of these completing things to where they are used.